src/shape_transform_descriptor.cpp

```diff
@@ -0,0 +1,840 @@
+            s.hidden_axis = nullopt;
+        }
+    }
+}
+
+static operation make_reshape_unsqueeze(const std::vector<dimension::sub>& subs)
+{
+    bool use_reshape = false;
+    // Check if split dimensions are all additional 1s
+    if(std::any_of(
+           subs.begin(), subs.end(), [](const dimension::sub& s) { return s.axis.size() > 1; }))
+    {
+        auto subs2   = subs;
+        auto by_axis = by(std::equal_to<>{}, [](const dimension::sub& s) -> int64_t {
+            if(s.axis.empty())
+                return -1;
+            return s.axis.front();
+        });
+        group_by(
+            subs2.begin(),
+            subs2.end(),
+            [&](auto start, auto last) {
+                if(use_reshape)
+                    return;
+                // Number of elements
+                auto n = std::distance(start, last);
+                if(n < 2)
+                    return;
+                // Number of elements that are 1
+                auto n1 =
+                    std::count_if(start, last, [](const dimension::sub& s) { return s.len == 1; });
+                use_reshape |= std::max<int64_t>(0, n - n1 - 1) > 0;
+            },
+            by_axis);
+    }
+    if(use_reshape)
+    {
+        std::vector<std::size_t> dims;
+        std::transform(subs.begin(),
+                       subs.end(),
+                       std::back_inserter(dims),
+                       [](const dimension::sub& s) -> std::size_t {
+                           if(s.axis.empty())
+                               return 1;
+                           return s.len;
+                       });
+        return make_op("reshape", {{"dims", dims}});
+    }
+    else
+    {
+        std::vector<std::size_t> axes;
+        for(auto i : range(subs.size()))
+        {
+            const auto& sub = subs[i];
+            if(sub.axis.size() == 1)
+                continue;
+            if(sub.len != 1 and not sub.axis.empty())
+                continue;
+            axes.push_back(i);
+        }
+        return make_op("unsqueeze", {{"axes", axes}});
+    }
+}
+
+static bool has_no_axes(const dimension& d)
+{
+    return std::all_of(d.subdimensions.begin(), d.subdimensions.end(), [](const dimension::sub& s) {
+        return s.axis.empty() and not s.hidden_axis.has_value();
+    });
+}
+static bool has_axes(const dimension& d)
+{
+    return std::any_of(d.subdimensions.begin(), d.subdimensions.end(), [](const dimension::sub& s) {
+        return not s.axis.empty();
+    });
+}
+
+// This will generate the operators to apply the shape transformation that is
+// represented by this class. This is the order of operators that will be
+// generated if needed:
+//
+// 1. Reshape/unsqueeze
+// 2. Transpose
+// 3. Broadcast
+// 4. Reshape/squeeze
+// 5. Broadcast
+//
+// This will generate operators backwrds starting at 5 and going up. Steps 1-3
+// are generated from the subdimesions and steps 4-5 are generated with the
+// dimensions.
```

Comment:
```suggestion
// This will generate operators backwards starting at 5 and going up. Steps 1-3
// are generated from the subdimensions and steps 4-5 are generated with the
// dimensions.
```